src/main/java/net/dries007/tfc/common/blocks/devices/TFCComposterBlock.java

```diff
@@ -0,0 +1,131 @@
+
+public class TFCComposterBlock extends ExtendedBlock implements EntityBlockExtension
+{
+    public static final IntegerProperty STAGE = TFCBlockStateProperties.STAGE_8;
+    public static final EnumProperty<CompostType> TYPE = TFCBlockStateProperties.COMPOST_TYPE;
+
+    public TFCComposterBlock(ExtendedProperties properties)
+    {
+        super(properties);
+        registerDefaultState(getStateDefinition().any().setValue(STAGE, 0).setValue(TYPE, CompostType.NORMAL));
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public InteractionResult use(BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult result)
+    {
+        return level.getBlockEntity(pos, TFCBlockEntities.COMPOSTER.get()).map(composter -> composter.use(player.getItemInHand(hand), player, level.isClientSide)).orElse(InteractionResult.PASS);
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public void randomTick(BlockState state, ServerLevel level, BlockPos pos, Random random)
+    {
+        level.getBlockEntity(pos, TFCBlockEntities.COMPOSTER.get()).ifPresent(ComposterBlockEntity::randomTick);
+    }
+
+    @Override
+    public void animateTick(BlockState state, Level level, BlockPos pos, Random random)
+    {
+        final CompostType type = state.getValue(TYPE);
+        if (type == CompostType.NORMAL) return;
+        SimpleParticleType particle = type == CompostType.READY ? TFCParticles.COMPOST_READY.get().getType() : TFCParticles.COMPOST_ROTTEN.get().getType();
+
+        final double x = pos.getX() + random.nextDouble();
+        final double y = pos.getY() + 1 + random.nextDouble() / 5D;
+        final double z = pos.getZ() + random.nextDouble();
+        final int count = Mth.nextInt(random, 0, 4);
+        for (int i = 0; i < count; i++)
+        {
+            level.addParticle(particle, x, y, z, 0D, 0D, 0D);
+        }
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public BlockState updateShape(BlockState state, Direction facing, BlockState facingState, LevelAccessor level, BlockPos currentPos, BlockPos facingPos)
+    {
+        return state.canSurvive(level, currentPos) ? state : Blocks.AIR.defaultBlockState();
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos)
+    {
+        return level.getBlockState(pos.below()).isFaceSturdy(level, pos, Direction.UP);
+    }
+
+    @Override
+    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder)
+    {
+        super.createBlockStateDefinition(builder.add(STAGE, TYPE));
+    }
+
+    @Override
+    public void setPlacedBy(Level level, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack stack)
+    {
+        level.getBlockEntity(pos, TFCBlockEntities.COMPOSTER.get()).ifPresent(TickCounterBlockEntity::resetCounter);
```

Comment:
Does this actually work? As per discussion on discord I thought `setPlacedBy` doesn't actually have the BE available?